### PR TITLE
[MODULAR] Anti-Gravity Harness

### PR DESCRIPTION
**Pull request by Nerev4r**
## About The Pull Request
_Gravity, is a harness... and I have harnessed the harness._

vHMJnkN

PR is ready! The original idea comes from bluwu016 in #6071 but this will add a gravity-harness that can be purchased from Cargo as a goodie. Modes right now are anti-gravity and guaranteed gravity™ like magboots. It has the same storage capacity as a roundstart MODsuit, so really not that much at all. It will cost ~1200 from Cargo as a goodie.

For balance, it is back-mounted and requires power cells to function. Sprites by @zydras. Yes, it is Teshari/digitigrade compatible.

## How This Contributes To The Skyrat Roleplay Experience
Given the upcoming existence of the Spacer Born quirk as well as paraplegic characters (and I would like, eventually, to add support for people with the Bad Back quirk), having a futuristic disability aid that can also function as a technology upgrade for miners/moths/anyone that benefits from anti-gravity, I thought it would be prudent to add in something for those crowds.

## Proof of Testing

<details>
<summary>Screenshots/Videos</summary>
  
vHMJnkN

</details>

## Changelog
:cl: Zydras
add: Purchasable gravity harness in the Cargo goodies section.
/:cl:

### CODE REVIEW

**Comment by Paxilmaniac** [comment]
> guaranteed gravity™

Using this mode in space should banish you to the void like using the atrocinator on lavaland does.

**Comment by W4rd3nn** [comment]
> > guaranteed gravity™
> 
> Using this mode in space should banish you to the void like using the atrocinator on lavaland does.

this as hell.

also very good pr maintainers speedmerge this NOW

File: modular_skyrat/master_files/code/modules/clothing/back/antigravityharness.dm
Comment (softcerv) [inline modular_skyrat/master_files/code/modules/clothing/back/antigravityharness.dm]: we need autodoc comments for these variables.

File: modular_skyrat/master_files/code/modules/clothing/back/antigravityharness.dm
Comment (softcerv) [inline modular_skyrat/master_files/code/modules/clothing/back/antigravityharness.dm]: We don't need to have this. You can modify to the `switch` to have it change the mode variable. PM me on discord if you need help.

File: modular_skyrat/master_files/code/modules/clothing/back/antigravityharness.dm
Comment (softcerv) [inline modular_skyrat/master_files/code/modules/clothing/back/antigravityharness.dm]: what does this do?

File: modular_skyrat/master_files/code/modules/clothing/back/antigravityharness.dm
Comment (softcerv) [inline modular_skyrat/master_files/code/modules/clothing/back/antigravityharness.dm]: These can probably be converted into defines

File: modular_skyrat/master_files/code/modules/clothing/back/antigravityharness.dm
Comment (softcerv) [inline modular_skyrat/master_files/code/modules/clothing/back/antigravityharness.dm]: This can be converted into a define too

File: modular_skyrat/master_files/code/modules/clothing/back/antigravityharness.dm
Comment (softcerv) [inline modular_skyrat/master_files/code/modules/clothing/back/antigravityharness.dm]:
```suggestion
	switch(mode)
		if(MODE_ANTIGRAVITY)
			if(user.has_gravity())
				new /obj/effect/temp_visual/mook_dust(get_turf(src))
			user.AddElement(/datum/element/forced_gravity, 0)
			playsound(src, 'sound/effects/gravhit.ogg', 50)
			to_chat(user, span_notice("Your harness releases a metallic hum, projecting a local anti-gravity field."))
			gravity_on = TRUE
			icon_state = antigravity_state
			worn_icon_state = antigravity_state
			
		if(MODE_EXTRAGRAVITY)
			if(!user.has_gravity())
				new /obj/effect/temp_visual/mook_dust/robot(get_turf(src))
			user.RemoveElement(/datum/element/forced_gravity, 0)
			ADD_TRAIT(user, TRAIT_NEGATES_GRAVITY)
			playsound(src, 'modular_skyrat/master_files/sound/effects/robot_sit.ogg', 25)
			to_chat(user, span_notice("Your harness shudders and hisses, projecting a local extra-gravity field."))
			gravity_on = TRUE
			worn_icon_state = extragravity_state
			
		if(MODE_GRAVOFF)
			if(!user.has_gravity())
				new /obj/effect/temp_visual/mook_dust/robot(get_turf(src))
				playsound(src, 'modular_skyrat/master_files/sound/effects/robot_sit.ogg', 25)
				to_chat(user, span_notice("Your harness lets out a soft whine as your gravity field dissipates, your body free-floating once again."))
			
			else
				if(user.has_gravity())
					new /obj/effect/temp_visual/mook_dust(get_turf(src))
					playsound(src, 'sound/effects/gravhit.ogg', 50)
					to_chat(user, span_notice("Your harness lets out a soft whine as your gravity field dissipates, leaving your body grounded once again."))
			user.RemoveElement(/datum/element/forced_gravity, 0)
			REMOVE_TRAIT(user, TRAIT_NEGATES_GRAVITY)
			icon_state = off_state
			worn_icon_state = off_state
			gravity_on = FALSE
```

File: modular_skyrat/master_files/code/modules/clothing/back/antigravityharness.dm
Comment (softcerv) [inline modular_skyrat/master_files/code/modules/clothing/back/antigravityharness.dm]:
```suggestion
	if(in_range(src,user) || isobserver(user))
		. += "The gravity harness is [gravity_on ? "on" : "off"] and the field is set to [mode]"
		. += "The power meter shows [cell ? "[round(cell.percent(), 0.1)]%" : "!invalid!"] charge remaining."
		if(cell_cover_open)
			. += "The cell cover is open, exposing the battery."
			if(!cell)
				. += "The cell slot is empty, showing bare connectors."
			else
				. += "\The [cell] is firmly in place."
		
	return .
```

File: modular_skyrat/master_files/code/modules/clothing/back/antigravityharness.dm
Comment (softcerv) [inline modular_skyrat/master_files/code/modules/clothing/back/antigravityharness.dm]: You could probably refactor this into being an early return.

File: modular_skyrat/master_files/code/modules/clothing/back/antigravityharness.dm
Comment (softcerv) [inline modular_skyrat/master_files/code/modules/clothing/back/antigravityharness.dm]: same here

Comment (softcerv) [inline modular_skyrat/master_files/code/modules/clothing/back/antigravityharness.dm]:
```suggestion
		return TRUE
		
	return ..()
```

File: modular_skyrat/master_files/code/modules/clothing/back/antigravityharness.dm
Comment (softcerv) [inline modular_skyrat/master_files/code/modules/clothing/back/antigravityharness.dm]:
```suggestion
```

File: modular_skyrat/master_files/code/modules/clothing/back/antigravityharness.dm
Comment (softcerv) [inline modular_skyrat/master_files/code/modules/clothing/back/antigravityharness.dm]: You have two of the same proc, I am not sure if you want to use this version of the proc or the one before.

File: modular_skyrat/master_files/code/modules/clothing/back/antigravityharness.dm
Comment (softcerv) [inline modular_skyrat/master_files/code/modules/clothing/back/antigravityharness.dm]:
```suggestion
		QDEL_NULL(cell)
		
```

**Comment by Elli-Skala** [comment]
Doesn't this undermind a mod suit module that requires a anomaly core and its whole tech just to make?

**Comment by softcerv** [comment]
> Doesn't this undermind a mod suit module that requires a anomaly core and its whole tech just to make?

This prevents you from using a MODSuit or backpack though. I feel like that is a reasonable enough trade-off.

**Review by softcerv — COMMENTED** [review]
some stuff I noticed, there will probably be more later.

File: modular_skyrat/master_files/code/modules/clothing/back/antigravityharness.dm
Comment (softcerv) [inline modular_skyrat/master_files/code/modules/clothing/back/antigravityharness.dm]:
```suggestion
	return
```

**Comment by Nerev4r** [comment]
undrafting tmrw in anticipation of spritework

File: modular_skyrat/master_files/code/modules/clothing/back/antigravityharness.dm
Comment (GoldenAlpharex) [inline modular_skyrat/master_files/code/modules/clothing/back/antigravityharness.dm]: Document this proc, pretty please.
```suggestion
/obj/item/gravityharness/proc/toggle_mode(mob/user, voluntary)
```

File: modular_skyrat/master_files/code/modules/clothing/back/antigravityharness.dm
Comment (GoldenAlpharex) [inline modular_skyrat/master_files/code/modules/clothing/back/antigravityharness.dm]:
```suggestion
///Changes the mode to `target_mode`, returns `FALSE` if the mode cannot be changed
```

File: modular_skyrat/master_files/code/modules/clothing/back/antigravityharness.dm
Comment (GoldenAlpharex) [inline modular_skyrat/master_files/code/modules/clothing/back/antigravityharness.dm:109]:
```suggestion

			if(user.has_gravity())
				new /obj/effect/temp_visual/mook_dust(get_turf(src))

```

File: modular_skyrat/master_files/code/modules/clothing/back/antigravityharness.dm
Comment (GoldenAlpharex) [inline modular_skyrat/master_files/code/modules/clothing/back/antigravityharness.dm:121]:
```suggestion

			if(!user.has_gravity())
```

File: modular_skyrat/master_files/code/modules/clothing/back/antigravityharness.dm
Comment (GoldenAlpharex) [inline modular_skyrat/master_files/code/modules/clothing/back/antigravityharness.dm]:
```suggestion
				to_chat(user, span_notice("[src] lets out a soft whine as your suspension field dissipates, gravity around you normalizing."))
```

File: modular_skyrat/master_files/code/modules/clothing/back/antigravityharness.dm
Comment (GoldenAlpharex) [inline modular_skyrat/master_files/code/modules/clothing/back/antigravityharness.dm]:
```suggestion
					to_chat(user, span_notice("[src] lets out a soft whine as your suspension field dissipates, gravity around you normalizing."))
```

File: modular_skyrat/master_files/code/modules/clothing/back/antigravityharness.dm
Comment (GoldenAlpharex) [inline modular_skyrat/master_files/code/modules/clothing/back/antigravityharness.dm]: Document this please.

File: modular_skyrat/master_files/code/modules/clothing/back/antigravityharness.dm
Comment (GoldenAlpharex) [inline modular_skyrat/master_files/code/modules/clothing/back/antigravityharness.dm]: Try to change the typepath into this everywhere, so it follows the `snake_case` conventions.
```suggestion
/obj/item/gravity_harness/Initialize(mapload)
```

File: modular_skyrat/master_files/code/modules/clothing/back/antigravityharness.dm
Comment (GoldenAlpharex) [inline modular_skyrat/master_files/code/modules/clothing/back/antigravityharness.dm]: By convention, we try to have the `Destroy()` proc at the start of the proc definitions of a new class, after `Initialize()`, so it's easy to validate.

File: modular_skyrat/master_files/code/modules/clothing/back/antigravityharness.dm
Comment (GoldenAlpharex) [inline modular_skyrat/master_files/code/modules/clothing/back/antigravityharness.dm]:
```suggestion
/obj/item/gravity_harness/handle_atom_del(atom/harness_cell)
	if(harness_cell == current_cell)
```

File: modular_skyrat/master_files/code/modules/clothing/back/antigravityharness.dm
Comment (GoldenAlpharex) [inline modular_skyrat/master_files/code/modules/clothing/back/antigravityharness.dm:200]: You also probably want to do `current_cell = null` here after the `change_mode` call, so we avoid hard-deletions.

File: modular_skyrat/master_files/code/modules/clothing/back/antigravityharness.dm
Comment (GoldenAlpharex) [inline modular_skyrat/master_files/code/modules/clothing/back/antigravityharness.dm]:
```suggestion
	if(in_range(src, user) || isobserver(user))
```

File: modular_skyrat/master_files/code/modules/clothing/back/antigravityharness.dm
Comment (GoldenAlpharex) [inline modular_skyrat/master_files/code/modules/clothing/back/antigravityharness.dm]: It could be cool to have this using `span_warning()`.

File: modular_skyrat/master_files/code/modules/clothing/back/antigravityharness.dm
Comment (GoldenAlpharex) [inline modular_skyrat/master_files/code/modules/clothing/back/antigravityharness.dm]:
```suggestion
		. += "The power meter shows [current_cell ? "<b>[round(current_cell.percent(), 0.1)]%</b> charge remaining." : "[span_warning("\"MISSING CELL\"")]"]"
```

File: modular_skyrat/modules/cargo/code/goodies.dm
Comment (GoldenAlpharex) [inline modular_skyrat/modules/cargo/code/goodies.dm]:
```suggestion
		While it's a pretty cheap bootleg of the personal gravitic engines used by the Skrell, this one's been fitted with straps and a basic storage module."
```

**Review by GoldenAlpharex — CHANGES_REQUESTED** [review]
I like this thing.

File: modular_skyrat/master_files/code/modules/clothing/back/antigravityharness.dm
Comment (GoldenAlpharex) [inline modular_skyrat/master_files/code/modules/clothing/back/antigravityharness.dm:59]:
```suggestion
		mode = MODE_GRAVOFF

```

**Review by GoldenAlpharex — CHANGES_REQUESTED** [review]
Just one tiny change, also you've got merge conflicts now it seems.

**Review by GoldenAlpharex — APPROVED** [review]
I reckon this should be good to go if the tests pass.

We can always change the price later if we feel like it's too cheap at the moment.

**Pull request merged**
